File: app/src/main/java/org/alie/paintgradientwork/view/BitmapGradientView.java
package org.alie.paintgradientwork.view;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Shader;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import org.alie.paintgradientwork.R;


/**
 * Created by Alie on 2019/3/17.
 * 类描述 位图渲染
 * 版本
 */
public class BitmapGradientView extends View {
    private static final String TAG = "BitmapGradientView2";
    private Paint mPaint;
    private Bitmap mBitmap;
    private int mWidth;
    private int mHeight;
    private int[] mColors = {Color.RED,Color.YELLOW,Color.BLUE,Color.GREEN};

    public BitmapGradientView(Context context) {
        super(context);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mBitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.xyjy2);
        mWidth = mBitmap.getWidth();
        mHeight = mBitmap.getHeight();
    }

    public BitmapGradientView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mBitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.xyjy2);
        mWidth = mBitmap.getWidth();
        mHeight = mBitmap.getHeight();
    }

    public BitmapGradientView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mBitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.xyjy2);
        mWidth = mBitmap.getWidth();
        mHeight = mBitmap.getHeight();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        Log.i(TAG, "width:" + mWidth + " height:" + mHeight);
        BitmapShader bitmapShader = new BitmapShader(mBitmap, Shader.TileMode.CLAMP, Shader.TileMode.MIRROR);
        mPaint.setShader(bitmapShader);
        canvas.drawRect(10, 10, mWidth + 10, mHeight * 2+10,mPaint);

    }
}

File: app/src/main/java/org/alie/paintgradientwork/kt/gradient/BitmapGradientView.kt
package org.alie.paintgradientwork.kt.gradient

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import org.alie.paintgradientwork.R

/**
 * Created by Administrator on 2020/8/8 0008.
 * 类描述 位图渲染的view
 * 版本
 */
class BitmapGradientView(context: Context, attributeSet: AttributeSet?, defStyle: Int) :
    View(context, attributeSet, defStyle) {
    var mBitmap: Bitmap? = null
    var mPaint: Paint? = null
    var mWidth = 0
    var mHeight = 0
    var bitmapShader: BitmapShader? = null

    init {
        mBitmap = BitmapFactory.decodeResource(resources, R.drawable.xyjy2)
        mWidth = mBitmap!!.width
        mHeight = mBitmap!!.height
        mPaint = Paint()
        // 关键代码，创建位图着色器，传入目标bitmap和多出bitmap宽高部分的拉伸模式
        bitmapShader = BitmapShader(mBitmap!!, Shader.TileMode.CLAMP, Shader.TileMode.MIRROR)
        mPaint?.let {
            it.isAntiAlias = true
            it.shader = bitmapShader
        }
    }

    constructor(context: Context, attributeSet: AttributeSet?) : this(context, attributeSet, 0)
    constructor(context: Context) : this(context, null, 0)


    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas?.drawRect(
            10F, 10F, (mWidth + 10).toFloat(), (mHeight + 10 + mHeight).toFloat(),
            mPaint!!
        )
    }
}